// @visheratin/web-ai-node/multimodal/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseMultimodalModel = void 0;
const image_js_1 = require("../image.js");
const index_js_1 = require("../index.js");
class BaseMultimodalModel {
    constructor(metadata) {
        this.init = async (proxy = true) => {
            const start = new Date();
            for (const [name, path] of this.metadata.modelPaths) {
                if (!this.sessions) {
                    this.sessions = new Map();
                }
                this.sessions.set(name, await (0, index_js_1.createSession)(path, proxy));
            }
            const preprocessorConfig = await image_js_1.PreprocessorConfig.fromFile(this.metadata.preprocessorPath);
            this.preprocessor = new image_js_1.Preprocessor(preprocessorConfig);
            this.tokenizer = await (0, index_js_1.loadTokenizer)(this.metadata.tokenizerPath);
            this.initialized = true;
            const end = new Date();
            const elapsed = (end.getTime() - start.getTime()) / 1000;
            return elapsed;
        };
        if (index_js_1.SessionParams.memoryLimitMB > 0 &&
            index_js_1.SessionParams.memoryLimitMB < metadata.memEstimateMB) {
            throw new Error(`The model requires ${metadata.memEstimateMB} MB of memory, but the current memory limit is 
          ${index_js_1.SessionParams.memoryLimitMB} MB.`);
        }
        this.metadata = metadata;
        this.initialized = false;
    }
}
exports.BaseMultimodalModel = BaseMultimodalModel;
