// @visheratin/web-ai-node/image/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseImageModel = void 0;
const preprocessor_js_1 = __importDefault(require("./preprocessor.js"));
const preprocessorConfig_js_1 = require("./preprocessorConfig.js");
const config_js_1 = __importDefault(require("./config.js"));
const index_js_1 = require("../index.js");
class BaseImageModel {
    constructor(metadata) {
        this.init = async (proxy = true) => {
            const start = new Date();
            for (const [name, path] of this.metadata.modelPaths) {
                if (!this.sessions) {
                    this.sessions = new Map();
                }
                this.sessions.set(name, await (0, index_js_1.createSession)(path, proxy));
            }
            const preprocessorConfig = await preprocessorConfig_js_1.PreprocessorConfig.fromFile(this.metadata.preprocessorPath);
            this.preprocessor = new preprocessor_js_1.default(preprocessorConfig);
            if (this.metadata.configPath) {
                this.config = await config_js_1.default.fromFile(this.metadata.configPath);
            }
            this.initialized = true;
            const end = new Date();
            const elapsed = (end.getTime() - start.getTime()) / 1000;
            return elapsed;
        };
        if (index_js_1.SessionParams.memoryLimitMB > 0 &&
            index_js_1.SessionParams.memoryLimitMB < metadata.memEstimateMB) {
            throw new Error(`The model requires ${metadata.memEstimateMB} MB of memory, but the current memory limit is 
          ${index_js_1.SessionParams.memoryLimitMB} MB.`);
        }
        this.metadata = metadata;
        this.initialized = false;
    }
}
exports.BaseImageModel = BaseImageModel;
