// @visheratin/web-ai-node/node/tokenizer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tokenizer = void 0;
const tokenizers_node_1 = require("@visheratin/tokenizers-node");
const node_fetch_1 = __importDefault(require("node-fetch"));
class Tokenizer {
    constructor(tokenizerPath) {
        this.tokenizerPath = tokenizerPath;
    }
    async init() {
        const response = await (0, node_fetch_1.default)(this.tokenizerPath);
        const tokenizerData = await response.json();
        tokenizerData["padding"] = null;
        this.instance = new tokenizers_node_1.WasmTokenizer(JSON.stringify(tokenizerData));
    }
    async decode(ids, skip_special_tokens) {
        if (this.instance === undefined) {
            throw new Error("Tokenizer is not initialized");
        }
        return this.instance.decode(ids, skip_special_tokens);
    }
    async encode(text, add_special_tokens) {
        if (this.instance === undefined) {
            throw new Error("Tokenizer is not initialized");
        }
        return this.instance.encode(text, add_special_tokens);
    }
}
exports.Tokenizer = Tokenizer;
